codec-http/src/main/java/io/netty/handler/codec/http/cors/CorsHandler.java

```diff
@@ -87,14 +88,35 @@ public void channelRead(final ChannelHandlerContext ctx, final Object msg) throw
                 handlePreflight(ctx, request);
                 return;
             }
-            if (isShortCircuit && !(origin == null || config != null)) {
+            if (isShortCircuit && !allowRequest(ctx, origin)) {
                 forbidden(ctx, request);
                 return;
             }
         }
         ctx.fireChannelRead(msg);
     }
 
+    private boolean allowRequest(ChannelHandlerContext ctx, String origin) {
+        if (origin == null) {
+            // Not a CORS Request, allow it to proceed.
+            return true;
+        }
+        if(config != null) {
+            // A rule was matched from the configs, so allow it.
+            return true;
+        }
+
+        String scheme = "http";
+        if(ctx.channel().pipeline().get(SslHandler.class) != null) {
+            scheme = "https";
+        }
+
+        String host = scheme + "://" + request.headers().get(HttpHeaderNames.HOST);
+
+        // If it's the same origin, allow it to proceed.
+        return origin.equals(host);
```

Comment:
do we need to ignore case ?

Reply:
I would say yes, since host names are case insensitive.